IonHashDotnet.Tests/IonHashTest.cs

```diff
@@ -0,0 +1,300 @@
+                {
+                    IIonValue i = bytesEnum.Current;
+                    if (cnt++ > 0)
+                    {
+                        sb.Append(" ");
+                    }
+
+                    sb.Append(i.IntValue.ToString("x2"));
+                }
+
+                sb.Append(")");
+            }
+
+            if (multipleEntries)
+            {
+                sb.Append("\n");
+            }
+
+            sb.Append(")");
+            return sb.ToString();
+        }
+
+        internal abstract class IonHashTester
+        {
+            private protected TestIonHasherProvider hasherProvider;
+
+            internal virtual IIonReader GetIonReader(string ionText)
+            {
+                return IonReaderBuilder.Build(ionText);
+            }
+
+            internal virtual IIonReader GetIonReader(byte[] ionBinary)
+            {
+                return IonReaderBuilder.Build(ionBinary);
+            }
+            internal virtual IIonReader GetIonReader(IIonValue ionValue)
+            {
+                return IonReaderBuilder.Build(ionValue);
+            }
+
+            internal virtual void Traverse(IIonReader reader, TestIonHasherProvider hasherProvider)
+            {
+                this.hasherProvider = hasherProvider;
+                IIonHashReader ihr = new IonHashReader(reader, hasherProvider);
+                Traverse(ihr);
+                ihr.Digest();
+            }
+
+            internal virtual void Traverse(IIonHashReader reader)
+            {
+                IonType iType;
+                while ((iType = reader.MoveNext()) != IonType.None)
+                {
+                    if (!reader.CurrentIsNull && IonTypeExtensions.IsContainer(iType))
+                    {
+                        reader.StepIn();
+                        Traverse(reader);
+                        reader.StepOut();
+                    }
+                }
+            }
+
+            internal virtual IIonValue GetHashLog()
+            {
+                return this.hasherProvider.GetHashLog();
+            }
+        }
+
+        internal class BinaryTest : IonHashTester
+        {
+            internal override IIonReader GetIonReader(string ionText)
+            {
+                MemoryStream ms = new MemoryStream();
+                IIonWriter writer = IonBinaryWriterBuilder.Build(ms);
+                IIonReader reader = IonReaderBuilder.Build(ionText);
+                writer.WriteValues(reader);
+                writer.Flush();
+                return IonReaderBuilder.Build(ms.ToArray());
+            }
+        }
+
+        internal class DomTest : IonHashTester { }
+
+        internal class TextTest : IonHashTester { }
+
+        internal class TextNoStepInTest : TextTest
+        {
+            internal override void Traverse(IIonHashReader reader)
+            {
+                while (reader.MoveNext() != IonType.None) { }
+            }
+        }
+
```

Comment:
Because there's no equivalent in C#.

More specifically, it would just be the BinaryTest, but instead of passing in ms.ToArray(), we would pass in new MemoryStream(ms.ToArray), which is redundant because the reader builder that takes in a byte[] does exactly the same thing, passing it to the overloaded build function by calling new MemoyStream on it.

Reply:
Looking at IonReaderBuilder...got it.  As long as other Stream implementations behave in the same manner as MemoryStream, I don't think we're missing any coverage here.  Thank you.